contrib/inventory/foreman.py

```diff
@@ -0,0 +1,394 @@
+        else:
+            d[k] = [v]
+
+    def run(self):
+        if not self._read_settings():
+            return False
+        self._get_inventory()
+        self._print_data()
+        return True
+
+    def _read_settings(self):
+        # Read settings and parse CLI arguments
+        if not self.read_settings():
+            return False
+        self.parse_cli_args()
+        return True
+
+    def _get_inventory(self):
+        if self.args.refresh_cache:
+            self.update_cache()
+        elif not self.is_cache_valid():
+            self.update_cache()
+        else:
+            self.load_inventory_from_cache()
+            self.load_params_from_cache()
+            self.load_facts_from_cache()
+            self.load_cache_from_cache()
+
+    def _print_data(self):
+        data_to_print = ""
+        if self.args.host:
+            data_to_print += self.get_host_info()
+        else:
+            self.inventory['_meta'] = {'hostvars': {}}
+            for hostname in self.cache:
+                self.inventory['_meta']['hostvars'][hostname] = {
+                    'foreman': self.cache[hostname],
+                    'foreman_params': self.params[hostname],
+                }
+                if self.want_facts:
+                    self.inventory['_meta']['hostvars'][hostname]['foreman_facts'] = self.facts[hostname]
+
+            data_to_print += self.json_format_dict(self.inventory, True)
+
+        print(data_to_print)
+
+    def is_cache_valid(self):
+        """Determines if the cache is still valid"""
+        if os.path.isfile(self.cache_path_cache):
+            mod_time = os.path.getmtime(self.cache_path_cache)
+            current_time = time()
+            if (mod_time + self.cache_max_age) > current_time:
+                if (os.path.isfile(self.cache_path_inventory) and
+                    os.path.isfile(self.cache_path_params) and
+                        os.path.isfile(self.cache_path_facts)):
+                    return True
+        return False
+
+    def read_settings(self):
+        """Reads the settings from the foreman.ini file"""
+
+        config = ConfigParser.SafeConfigParser()
+        config.read(self.config_paths)
+
+        # Foreman API related
+        try:
+            self.foreman_url = config.get('foreman', 'url')
+            self.foreman_user = config.get('foreman', 'user')
+            self.foreman_pw = config.get('foreman', 'password')
+            self.foreman_ssl_verify = config.getboolean('foreman', 'ssl_verify')
+        except (ConfigParser.NoOptionError, ConfigParser.NoSectionError) as e:
+            print("Error parsing configuration: %s" % e, file=sys.stderr)
+            return False
+
+        # Ansible related
+        try:
+            group_patterns = config.get('ansible', 'group_patterns')
+        except (ConfigParser.NoOptionError, ConfigParser.NoSectionError):
+            group_patterns = "[]"
+
+        self.group_patterns = eval(group_patterns)
```

Comment:
I didn't notice this before but you have eval here... What is that for?  It's rather dangerous as it means if someone can get code into the config file for the foreman dynamic inventory, they can then execute whatever code they want as the user running ansible.  We should get rid of this if at all possible.